### PR TITLE
 Add pooling for JIT scratch memory 

### LINKED ISSUE #3408: JIT's allocator performance seems to be low
I profiled `crossgen mscorlib.dll` out of curiosity and noticed that ~5% of the time is spent in `VirtualFree`. These free calls come from the `ArenaAllocator` (formerly known as `norls_allocator`). It's kind of funny that freeing memory takes about the same time as other JIT actions that do actual work (for example, `SsaBuilder::Build` takes ~5% as well).

I suspect that this happens because the allocator's page is 64 kbytes and as a result a lot of pages are allocated and a lot of free calls have to be made. Though it's peculiar that allocation is very fast (time spent in `VirtualAlloc` is less than 1%), perhaps this is a Windows kernel issue.

In any case, 64 kbytes seems too low for 64 bit architectures. I tried with 1 megabyte and `crossgen mscorlib.dll` time drops down from ~6.8 seconds to ~6.0 seconds.

@pgavlin Perhaps you'll find this interesting, as far as I understand from your recent commit you're planning to do some more work around the allocator.

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #3408

### CODE REVIEW

**Comment by jkotas** [comment]
Perf results collected by Andy: https://github.com/dotnet/coreclr/issues/3408#issuecomment-404284071

Credit for this change also goes to @AndyAyersMS and @mikedn . They have run a lot of experiments that help us pick the current approach.

**Comment by mikedn** [comment]
This removes the only use of `ICorJitInfo::getMemoryManager()`. Presumably removing it from `ICorJitInfo` would be too much trouble for no gain?

File: src/jit/alloc.h [hunk 1]
```diff
@@ -47,28 +43,20 @@ class ArenaAllocator
     // Thus if we want to make this smaller, we need to do a reserve / commit scheme
     enum
     {
-        DEFAULT_PAGE_SIZE = 16 * OS_page_size,
+        DEFAULT_PAGE_SIZE = 0x10000,
```
Comment (mikedn) [inline src/jit/alloc.h]: The comment above enum has become stale since VirtualAlloc is no longer used

File: src/vm/jithost.cpp [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+void JitHost::Init()
+{
+    s_JitSlabAllocatorCrst.Init(CrstLeafLock);
+}
+
+void JitHost::Reclaim()
+{
+    if (s_pCurrentCachedList != NULL || s_pPreviousCachedList != NULL)
+    {
+        DWORD ticks = ::GetTickCount();
```
Comment (mikedn) [inline src/vm/jithost.cpp]: `GetTickCount64`?

File: src/vm/jithost.cpp [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+//
+
+struct Slab
+{
+    Slab * pNext;
+    size_t size;
+    Thread* affinity;
+};
+
+static CrstStatic s_JitSlabAllocatorCrst;
```
Comment (mikedn) [inline src/vm/jithost.cpp]: Any reason why these aren't member of `JitHost`?

File: src/vm/jithost.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+            s_TotalCached -= p->size;
+            *pActualSize = p->size;
+
+            return p;
+        }
+    }
+
+    *pActualSize = size;
+    return new (nothrow) BYTE[size];
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/vm/jithost.cpp]: It is our custom overload of operator new that is guaranteed do that. Anyway, I have switched it to use the lower-level malloc-like methods for clarity.

File: src/vm/jithost.cpp (hunk 2 above)
Comment (jkotas) [reply src/vm/jithost.cpp]: GetTickCount64 is slower on 32-bit. I think GetTickCount works just fine here.

File: src/vm/jithost.cpp (hunk 3 above)
Comment (jkotas) [reply src/vm/jithost.cpp]: Fixed

**Comment by jkotas** [comment]
OSX hit https://github.com/dotnet/coreclr/issues/18858

**Comment by jkotas** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test please

**Comment by mikedn** [comment]
In case you didn't already notice: it looks like this change will break CoreRT, its `JitHost` doesn't inherit from `ICorJitHost` so it won't get the default slab functions.

https://github.com/dotnet/corert/blob/master/src/Native/jitinterface/jithost.cpp

**Comment by jkotas** [comment]
cc @dotnet/jit-contrib

**Comment by jkotas** [comment]
(I will fix desktop once it gets there.)

**Comment by briansull** [comment]
This change has now reached the desktop

**Pull request merged**